### ISSUE TITLE
Allow more options for meta config format/name

### ISSUE THREAD

**Issue by gakada** [mateodelnorte/meta#149]
That is, to [allow](https://github.com/davidtheclark/cosmiconfig) not just `.meta` as a config file, but also `.meta.json`, `.meta.yaml`, `meta` property in `package.json`, etc.

**Comment by gakada** [comment]
On a side note, https://github.com/mateodelnorte/meta-init/blob/master/bin/meta-init and https://github.com/mateodelnorte/meta-project/blob/master/bin/meta-project-add format `.meta` without the final newline, so maybe https://github.com/mateodelnorte/get-meta-file should also encapsulate the formatting as an exported function (like `meta => JSON.stringify(meta, null, 2) + '\n'`), to use it in plugins.

**Comment by mateodelnorte** [comment]
Can you open a second issue for the second comment? 

On your first comment, what do you see as the added benefit? Seems like more maintenance footprint.

**Comment by gakada** [comment]
I want to treat `.meta` as a normal configuration file (that is, manually created, edited, checked in git), similar to [prettier config](https://prettier.io/docs/en/configuration.html) (which uses cosmiconfig), [eslint config](https://eslint.org/docs/user-guide/configuring#configuration-file-formats) (which is still similar to cosmiconfig), etc. And so then it depends on a project how those are presented, some prefer extensionless names, some names with extensions (e.g., my VS Code doesn't recognize `.meta` as a JSON file without extra configuration), some as `package.json` property.

So basically out of

- `name` in `package.json`
- `.name` (JSON or YAML)
- `.name.json`
- `.name.yaml`
- `.name.yml`
- `.name.js`
- `name.config.js`

`.meta` JSON is one particular way of storing the config, which may not match the conventions of a target repo while using cosmiconfig allows users to choose which one to use, based on their conventions.

**Comment by gakada** [comment]
Added https://github.com/mateodelnorte/get-meta-file/pull/22 for the second comment.

**Comment by mateodelnorte** [comment]
.meta should already be checked into git. If you want to create it manually, that's totally fine. That said, I'm fine following the lead of these projects. Want to make another PR in `get-meta-file` for that? Would also be nice to have a README update here, as well as in `get-meta-file`. 

I'd like to figure out an easy way to get the README files of each plugin to be a part of the overall `meta` docs. I took a swack at using gitbook for that, but it didn't work well due to some inane .gitignore stuff gitbook does.

**Comment by patrickleet** [comment]
I gave up on gitbook - they dropped support for legacy version and all their docs are 404s now.

I started using "bookdown" which is an R project, but specifically for making a PDF book.

I've heard good things about [Docusaurus](https://docusaurus.io/)

**Comment by gakada** [comment]
I will take a look. Right now it seems like `getFileLocation` and `format` might become redundant, leaving only `load` (default export one) and `save` (maybe with optional path), because with cosmiconfig it can be YAML, package.json, etc., something plugins should not handle themselves, only requiring `load`/`save` to/from an object (from/to default or custom path).

**Comment by gakada** [comment]
Well, some plugins need to change the config and cosmiconfig doesn't support saving, only loading, so I'm changing this :sweat_smile:

Still, I think there can be a relatively simple patch for `get-meta-file` that will allow more formats/names for the config. And it still stands that `get-meta-file` can provide just `load` and `save` and that plugins can interface with the config using only those.

The question is then what names to allow.

- Can have more JSON names, `.meta` as JSON, `.meta.json` (explicit extension for editors), `.metarc` as JSON and `.metarc.json` (ala babel, eslint, prettier, jest), `meta.json` (ala tslint), `meta.config.json` (?).
- YAML should be possible too, not sure if really needed though.
- Probably no `meta` in `package.json` support, since it will conflict with npm/yarn formatters (which behave differently from `JSON.stringify` IINM).
- Also not sure about JS, e.g., currently it can be JS in `.meta`, but `project add` will reformat it as JSON. Maybe JS config support should be deprecated since generally it is not possible to reformat it.

**Comment by mateodelnorte** [comment]
`meta` already supports a `.meta` file in both json and .js formats. If you use .js, you can put additional javascript behavior in it. Correct that `meta project add` will overwrite.  Absolutely not deprecating the js support, as there are companies that use it – especially in odd cases such as having to use a corporate hosted version of Bitbucket where you aren't allowed to use `git+ssh` and you have to get the local user's username to determine the `https` route for each repo.

**Comment by gakada** [comment]
Yeah, `.meta` in JSON or JS is the current format, so it is about adding lookup for other names and `meta` property in `package.json` (still planning to do a PR eventually btw).

JS configs are fine and practiced, if it is read-only configs, but say I have

```js
// some arbitrary JS code

module.exports = {
  // more JS code, using previous JS code, too
}
```

here it is not read-only, so `meta project add` will make it

```json
{
  some JSON after _evaluating_ the module
}
```

which may not be desired (like JS code and formatting are discarded). Also, if that JS is in `.meta.js`/`meta.config.js`/etc. (explicit extension), then it is even more complicated (`meta project add` should rename the file since it is not JS anymore?).

**Comment by mateodelnorte** [comment]
You can always edit your .meta file manually, and if you're using javascript in your .meta file that's what you'll do. The more code we write, the more we have to maintain. I'm not convinced supporting all these file types with add/remove of projects lends a lot of productivity. Being able to make git branches across many projects at once, see the git status of everything – those increase your productivity. 

I would argue that if we support reading from all the different file types, we shouldn't support writing to any of them – but we already have the ability to read json and js, so we should just keep it.

**Comment by mateodelnorte** [comment]
Btw, it's more code to maintain, more code to document, and more documentation to maintain. 

Simpler is better.

**Comment by gakada** [comment]
Oh, right, if `.meta` is in JS, then the solution is not to use altering commands like `meta project add`, so that must be fine, though a bit inconsistent.

As for file types, my two reasons were

- Somehow VS Code doesn't highlight `.meta` files by default, JSON or JS, hence the "need" for extensions.
- All other configs might already be in a certain format (`package.json` properties or certain names/extensions), so it is not really about productivity, but unification with other tooling.

**Comment by mateodelnorte** [comment]
For syntax highlighting, just click the change language mode button on the bottom right:

![image](https://user-images.githubusercontent.com/554847/57791725-ff659c80-7702-11e9-9cf1-a312aeb1ee8f.png)

or CMD+Shift+P and select change language mode: 

![image](https://user-images.githubusercontent.com/554847/57791753-13a99980-7703-11e9-8045-c19faf936999.png)

There are plenty of other projects that have non-json and non-js extensions, causing them to not have highlighting. This is not a big deal. If you want highlighting for .meta files, a better idea would be to start a `meta` VS Code extension. We could put some really powerful commands into it, and it would cover your highlighting. 

I don't see a benefit to unifying with other tooling. You're creating a new repo for your meta repo. You don't have to have a package.json in your meta repo. You only would if you want to locally install plugins. And I don't see a reason to shove all your child project repo locations into your package.json. 

Shoving many things into fewer files === more merge conflicts. Separate, and simple is better.

**Comment by panoply** [comment]
FYI: You also set file associations in vscode, eg: 

```json
{  
 "files.associations": {
    ".meta": "json"
  }
}
```

**Comment by panoply** [comment]
FYI: You can also set file associations in vscode, eg: 

```json
{  
 "files.associations": {
    ".meta": "json"
  }
}
```
